File: src/main/java/top/fpsmaster/ui/custom/Component.java
package top.fpsmaster.ui.custom;

import top.fpsmaster.utils.render.draw.Rects;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.ScaledResolution;
import org.lwjgl.opengl.GL11;
import top.fpsmaster.FPSMaster;
import top.fpsmaster.features.impl.InterfaceModule;
import top.fpsmaster.features.manager.Category;
import top.fpsmaster.features.manager.Module;
import top.fpsmaster.font.impl.UFontRenderer;
import top.fpsmaster.modules.logger.ClientLogger;

import java.awt.*;

public class Component {
    private static final Color STENCIL_MASK_COLOR = new Color(255, 255, 255, 255);
    public static final float MIN_SCALE = 0.5f;
    public static final float MAX_SCALE = 4.5f;

    public InterfaceModule mod;

    public float x = 0f;

    public float y = 0f;

    public float width = 0f;

    public float height = 0f;

    public float scale = 1f;

    public boolean allowScale = false;

    /** Consecutive render failures; reset on any successful frame. See ComponentsManager. */
    public int renderFailures = 0;

    public Position position = Position.LT;

    @SuppressWarnings("unchecked")
    public Component(Class<?> clazz) {
        Module module;
        try {
            module = FPSMaster.moduleManager.getModule((Class<? extends Module>) clazz);
        } catch (IllegalStateException exception) {
            ClientLogger.warn("Missing interface module for component: " + clazz.getName());
            this.mod = new InterfaceModule(clazz.getSimpleName(), Category.Interface);
            this.mod.set(false);
            return;
        }
        if (module instanceof InterfaceModule) {
            this.mod = (InterfaceModule) module;
            return;
        }

        ClientLogger.warn("Missing interface module for component: " + clazz.getName());
        this.mod = new InterfaceModule(clazz.getSimpleName(), Category.Interface);
        this.mod.set(false);
    }

    public void draw(float x, float y) {
    }

    /**
     * Computes {@link #width}/{@link #height} for this frame, before anything reads them.
     *
     * <p>Historically every component assigned its size inside {@code draw()}, but anchoring
     * ({@link #getRealPosition}), hover testing, drag clamping and the blur mask all run <em>before</em>
     * {@code draw()} — so they used the previous frame's size. A component whose size changes each
     * frame (FPS going {@code 99fps} → {@code 100fps}, a potion expiring) visibly jitters, and on its
     * very first frame {@code width} is still 0.
     *
     * <p>Default implementation does nothing, so components that have not been migrated keep their old
     * behaviour. Override it, move the sizing math here, and {@code draw()} becomes pure rendering.
     */
    public void measure() {
    }

    /** Computes editor geometry using representative data when live data is absent. */
    public void measurePreview() {
        measure();
    }

    /** Draws the editor representation without mutating player or world state. */
    public void drawPreview(float x, float y) {
        draw(x, y);
    }

    /** Receives the rectangles that make up a component's background. */
    public interface ShapeSink {
        /**
         * @param x absolute left edge, already scaled by the caller (same convention as
         *          {@link #drawRect}: positions are the caller's job, sizes are the base class's)
         * @param y absolute top edge
         * @param width  logical width; the base class multiplies by {@link #scale}
         * @param height logical height
         */
        void rect(float x, float y, float width, float height);
    }

    /**
     * Declares the geometry of this component's background so the blur mask can reproduce it exactly
     * instead of guessing.
     *
     * <p>The mask used to assume every component draws one rectangle at {@code (rX - 2, rY)} sized
     * {@code width × height}. That holds for about two thirds of them; the rest either use a different
     * origin (Keystrokes, PotionDisplay) or draw several disjoint boxes (ArmorDisplay), so the blur
     * leaked into the gaps or sat a couple of pixels off.
     *
     * <p>Whatever a component declares here must match what it actually paints in {@code draw()}.
     */
    public void backgroundShape(ShapeSink sink, float originX, float originY) {
        sink.rect(originX - 2f, originY, width, height);
    }

    public float alpha = 0f;

    public boolean shouldDisplay() {
        return mod.isEnabled();
    }

    /**
     * Whether this component currently has geometry another component may align to.
     *
     * <p>Some enabled HUD modules are conditional: a target HUD has no box without a target, for
     * example. Those components override this rather than letting their last non-zero width/height
     * become an invisible snap target.
     */
    public boolean isVisibleForAlignment() {
        return shouldDisplay() && width > 0f && height > 0f;
    }

    public float[] getRealPosition() {
        return getRealPosition(new ScaledResolution(Minecraft.getMinecraft()));
    }

    public float[] getRealPosition(ScaledResolution sr) {
        float rX = 0f;
        float rY = 0f;
        x = Math.max(0f, Math.min(1f, x));
        y = Math.max(0f, Math.min(1f, y));

        // The space these coordinates live in is set by ComponentsManager, which converts the mouse
        // by sr.getScaleFactor() and has GuiScale scale the matrix by 2 / sr.getScaleFactor(). This
        // has to read the same factor. Reading the client's own UI scale instead - which is 1 unless
        // the interface is set to follow the game's - made the usable area half as wide and half as
        // tall, so a component could only be dragged around the top-left quarter of the screen.
        float scaleFactor = sr.getScaleFactor();
        float guiWidth = sr.getScaledWidth() / 2f * scaleFactor;
        float guiHeight = sr.getScaledHeight() / 2f * scaleFactor;

        // Anchors offset by the component's *rendered* size. width/height are logical units; everything
        // that touches the drawn box (drawRect, hover, drag clamping, blur mask) multiplies by scale,
        // so these must too — otherwise a scaled-up right-anchored component overflows the screen edge
        // by width * (scale - 1).
        float scaledWidth = width * scale;
        float scaledHeight = height * scale;

        switch (position) {
            case LT:
                rX = x * guiWidth / 2f;
                rY = y * guiHeight / 2f;
                break;
            case RT:
                rX = guiWidth - (x * guiWidth / 2f + scaledWidth);
                rY = y * guiHeight / 2f;
                break;
            case LB:
                rX = x * guiWidth / 2f;
                rY = guiHeight - (y * guiHeight / 2f + scaledHeight);
                break;
            case RB:
                rX = guiWidth - (x * guiWidth / 2f + scaledWidth);
                rY = guiHeight - (y * guiHeight / 2f + scaledHeight);
                break;
            case CT:
                rX = guiWidth / 2f - scaledWidth / 2f;
                rY = y * guiHeight / 2f;
                break;
        }
        return new float[]{rX, rY};
    }

    /** Converts a shared editor placement back into Edge's anchored config model. */
    public void setRealPosition(float realX, float realY, float guiWidth, float guiHeight, float newScale) {
        scale = Math.max(MIN_SCALE, Math.min(MAX_SCALE, newScale));
        float scaledWidth = width * scale;
        float scaledHeight = height * scale;
        boolean right = realX + scaledWidth / 2f > guiWidth / 2f;
        boolean bottom = realY + scaledHeight / 2f > guiHeight / 2f;
        position = right ? (bottom ? Position.RB : Position.RT) : (bottom ? Position.LB : Position.LT);
        x = right ? 2f * (guiWidth - realX - scaledWidth) / guiWidth : 2f * realX / guiWidth;
        y = bottom ? 2f * (guiHeight - realY - scaledHeight) / guiHeight : 2f * realY / guiHeight;
        x = Math.max(0f, Math.min(1f, x));
        y = Math.max(0f, Math.min(1f, y));
    }

    public void drawBlurMask(ScaledResolution sr) {
        if (!hasBackground() || width <= 0f || height <= 0f) {
            return;
        }
        float[] pos = getRealPosition(sr);
        boolean round = mod.rounded.getValue();
        int radius = mod.roundRadius.getValue().intValue();
        // Same geometry the component paints, so the mask can never drift from the background.
        backgroundShape((rectX, rectY, rectWidth, rectHeight) -> {
            float scaledWidth = rectWidth * scale;
            float scaledHeight = rectHeight * scale;
            if (scaledWidth <= 0f || scaledHeight <= 0f) {
                return;
            }
            if (round) {
                Rects.roundedImage(Math.round(rectX), Math.round(rectY), Math.round(scaledWidth), Math.round(scaledHeight), radius, STENCIL_MASK_COLOR);
            } else {
                Rects.fill(rectX, rectY, scaledWidth, scaledHeight, STENCIL_MASK_COLOR);
            }
        }, pos[0], pos[1]);
    }

    public void display(ScaledResolution sr, int mouseX, int mouseY) {
        float[] pos = getRealPosition(sr);
        draw(pos[0], pos[1]);
    }

    /**
     * A module only has a background if its traits say so. Checking {@code bg} alone is not enough:
     * the field exists on every InterfaceModule and defaults to {@code true}, so a text-only module
     * such as Sprint would otherwise have a blur mask stamped for a panel it never draws.
     */
    public boolean hasBackground() {
        return mod.has(InterfaceModule.Trait.BACKGROUND) && mod.bg.getValue();
    }

    public void drawRect(float x, float y, float width, float height, Color color) {
        float scaledWidth = width * scale;
        float scaledHeight = height * scale;

        if (hasBackground()) {
            if (mod.rounded.getValue()) {
                Rects.roundedImage(Math.round(x), Math.round(y), Math.round(scaledWidth), Math.round(scaledHeight), mod.roundRadius.getValue().intValue(), color);
            } else {
                Rects.fill(x, y, scaledWidth, scaledHeight, color);
            }
        }
    }

    public void drawString(int fontSize, String text, float x, float y, int color) {
        drawString(fontSize, false, text, x, y, color);
    }

    /**
     * Draws a component's text.
     *
     * <p>Colours here come from {@link top.fpsmaster.features.settings.impl.ColorSetting}, whose
     * alpha slider reaches zero and whose Wave mode scales what it returns, so alpha means what it
     * says: below four there is nothing to draw. Both renderers underneath read it vanilla's way
     * instead — no alpha bits set means opaque — and would turn a string the user hid into a solid
     * one. Components must therefore pass a real alpha, not a bare {@code 0xRRGGBB}.
     *
     * <p>Only text is affected. Backgrounds and shapes go through {@code drawRect} and {@code Rects}
     * and never come through here, so a fully transparent background stays transparent.
     */
    public void drawString(int fontSize, boolean bold, String text, float x, float y, int color) {
        if (((color >>> 24) & 0xFF) <= 3) {
            return;
        }
        double scaled = (int) (scale * 100) / 100.0;
        fontSize = (int) (fontSize * scale);
        UFontRenderer font = FPSMaster.fontManager.getFont(fontSize);
        if (mod.betterFont.getValue()) {
            if (mod.fontShadow.getValue()) font.drawStringWithShadow(text, x, y, color);
            else font.drawString(text, x, y, color);
        } else {
            GL11.glPushMatrix();
            GL11.glTranslated(x, y, 0.0);
            GL11.glScaled(scaled, scaled, 1.0);
                if (mod.fontShadow.getValue()) {
                    Minecraft.getMinecraft().fontRendererObj.drawStringWithShadow(text, 0, 0, color);
                } else {
                    GL11.glColor4f(1, 1, 1, 1);
                    Minecraft.getMinecraft().fontRendererObj.drawString(text, 0, 0, color);
                }
            GL11.glPopMatrix();
        }
    }

    public float getStringWidth(int fontSize, String name) {
        UFontRenderer font = FPSMaster.fontManager.getFont(fontSize);
        return mod.betterFont.getValue() ? font.getStringWidth(name) : (Minecraft.getMinecraft().fontRendererObj.getStringWidth(name));
    }

    public float getStringHeight(int fontSize) {
        UFontRenderer font = FPSMaster.fontManager.getFont(fontSize);
        return mod.betterFont.getValue() ? font.getHeight() : (Minecraft.getMinecraft().fontRendererObj.FONT_HEIGHT);
    }

}

File: shared/java/top/fpsmaster/ui/custom/Component.kt
package top.fpsmaster.ui.custom

import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiChat
import net.minecraft.client.gui.ScaledResolution
import org.lwjgl.input.Mouse
import org.lwjgl.opengl.GL11
import top.fpsmaster.FPSMaster
import top.fpsmaster.font.impl.UFontRenderer
import top.fpsmaster.features.impl.InterfaceModule
import top.fpsmaster.features.impl.interfaces.ClientSettings
import top.fpsmaster.ui.click.MainPanel
import top.fpsmaster.utils.Utility
import top.fpsmaster.utils.math.animation.AnimationUtils.base
import top.fpsmaster.utils.render.Render2DUtils
import top.fpsmaster.interfaces.ProviderManager
import top.fpsmaster.utils.render.shader.BlurBuffer
import java.awt.Color
import kotlin.math.max
import kotlin.math.min

open class Component(clazz: Class<*>?) {

    private var dragX = 0f
    private var dragY = 0f

    @JvmField
    var mod: InterfaceModule = FPSMaster.moduleManager.getModule(clazz!!) as InterfaceModule
    var position = Position.LT

    @JvmField
    var x = 0f

    @JvmField
    var y = 0f

    @JvmField
    var width = 0f

    @JvmField
    var height = 0f

    @JvmField
    var scale = 1f

    var allowScale = false


    open fun draw(x: Float, y: Float) {}
    var alpha = 0f

    fun shouldDisplay(): Boolean {
        return mod.isEnabled
    }

    fun getRealPosition(): FloatArray {
        val sr = ScaledResolution(Minecraft.getMinecraft())
        var rX = 0f
        var rY = 0f
        x = max(0f, min(1f, x))
        y = max(0f, min(1f, y))
        var scaleFactor: Int = if (ClientSettings.Companion.fixedScale.value) {
            sr.scaleFactor;
        } else {
            2;
        }
        val guiWidth = sr.scaledWidth / 2f * scaleFactor
        val guiHeight = sr.scaledHeight / 2f * scaleFactor

        when (position) {
            Position.LT -> {
                rX = x * guiWidth / 2f
                rY = y * guiHeight / 2f
            }

            Position.RT -> {
                rX = guiWidth - (x * guiWidth / 2f + width)
                rY = y * guiHeight / 2f
            }

            Position.LB -> {
                rX = x * guiWidth / 2f
                rY = guiHeight - (y * guiHeight / 2f + height)
            }

            Position.RB -> {
                rX = guiWidth - (x * guiWidth / 2f + width)
                rY = guiHeight - (y * guiHeight / 2f + height)
            }

            Position.CT -> {

            }
        }
        return floatArrayOf(rX, rY)
    }

    fun display(mouseX: Int, mouseY: Int) {
        val rX = getRealPosition()[0]
        val rY = getRealPosition()[1]
        draw(rX.toInt().toFloat(), rY.toInt().toFloat())
        if (Utility.mc.currentScreen !is GuiChat && Utility.mc.currentScreen !is MainPanel
        ) return
        var width = width * scale
        var height = height * scale
        val drag = FPSMaster.componentsManager.dragLock == mod.name
        alpha = if (Render2DUtils.isHovered(rX, rY, width, height, mouseX, mouseY) || drag) {
            if ((base(alpha.toDouble(), 50.0, 0.1).toFloat()).isNaN())
                0f else base(alpha.toDouble(), 50.0, 0.1).toFloat()
        } else {
            if ((base(alpha.toDouble(), 0.0, 0.1).toFloat()).isNaN())
                0f else base(alpha.toDouble(), 0.0, 0.1).toFloat()
        }
        Render2DUtils.drawOptimizedRoundedRect(rX - 2, rY - 2, width + 4, height + 4, Color(0, 0, 0, alpha.toInt()))
        if (!Mouse.isButtonDown(0)) {
            FPSMaster.componentsManager.dragLock = ""
        }
        if (Render2DUtils.isHovered(rX, rY, width, height, mouseX, mouseY) || drag) {
            if (allowScale) {
                val dWheel = Mouse.getDWheel()
                if (dWheel > 0) {
                    this.scaleUp()
                } else if (dWheel < 0) {
                    this.scaleDown()
                }
            }
            FPSMaster.fontManager.s14.drawString(
                FPSMaster.i18n[mod.name.lowercase()] + " " + (scale * 10).toInt() / 10f + "x",
                rX,
                rY - 10,
                -1
            )
            if (!Mouse.isButtonDown(0)) return
            if (!drag && FPSMaster.componentsManager.dragLock.isEmpty()) {
                dragX = mouseX - rX
                dragY = mouseY - rY
                FPSMaster.componentsManager.dragLock = mod.name
            }
            if (FPSMaster.componentsManager.dragLock == mod.name) {
                move(mouseX.toFloat(), mouseY.toFloat())
                FPSMaster.componentsManager.dragLock = mod.name
            }

        }


    }

    fun scaleUp() {
        if (scale < 2.5f)
            scale += 0.1f
    }

    fun scaleDown() {
        if (scale > 0.5f)
            scale -= 0.1f
    }

    private fun move(x: Float, y: Float) {
        val sr = ScaledResolution(Utility.mc)
        var scaleFactor: Int = if (ClientSettings.Companion.fixedScale.value) {
            sr.scaleFactor;
        } else {
            2;
        }
        val guiWidth = sr.scaledWidth / 2f * scaleFactor
        val guiHeight = sr.scaledHeight / 2f * scaleFactor
        var changeX = 0f
        var changeY = 0f
        if (x > guiWidth / 2f) {
            if (y >= guiHeight / 2f) position = Position.RB else if (y < guiHeight / 2f) position =
                Position.RT
        } else {
            if (y >= guiHeight / 2f) position = Position.LB else if (y < guiHeight / 2f) position =
                Position.LT
        }
        when (position) {
            Position.LT -> {
                changeX = x - dragX
                changeY = y - dragY
            }

            Position.RT -> {
                changeX = guiWidth - x - width + dragX
                changeY = y - dragY
            }

            Position.LB -> {
                changeX = x - dragX
                changeY = guiHeight - y - height + dragY
            }

            Position.RB -> {
                changeX = guiWidth - x - width + dragX
                changeY = guiHeight - y - height + dragY
            }

            Position.CT -> {

            }
        }
        this.x = changeX / guiWidth * 2f
        this.y = changeY / guiHeight * 2f
    }

    fun drawRect(x: Float, y: Float, width: Float, height: Float, color: Color?) {
        var width = width * scale
        var height = height * scale
        BlurBuffer.blurArea(x, y, width, height, true)
        if (mod.bg.value)
            if (mod.rounded.value) {
                Render2DUtils.drawOptimizedRoundedRect(x, y, width, height, mod.roundRadius.value.toInt(), color!!.rgb)
            } else {
                Render2DUtils.drawRect(x, y, width, height, color)
            }
    }

    fun drawString(fontSize: Int, text: String, x: Float, y: Float, color: Int) {
        drawString(fontSize, false, text, x, y, color)
    }

    fun drawString(fontSize: Int, bold: Boolean, text: String, x: Float, y: Float, color: Int) {
        val fontSize = (fontSize * scale).toInt()
        val font = FPSMaster.fontManager.getFont(fontSize)
        if (mod.betterFont.value) {
            if (mod.fontShadow.value) font.drawStringWithShadow(text, x, y, color) else font.drawString(
                text,
                x,
                y,
                color
            )
        } else {
            GL11.glPushMatrix()
            GL11.glTranslated(x.toDouble(), y.toDouble(), 0.0)
            GL11.glScaled(scale.toDouble(), scale.toDouble(), 1.0)
            if (mod.fontShadow.value) ProviderManager.mcProvider.getFontRenderer()
                .drawStringWithShadow(text, 0f, 0f, color) else ProviderManager.mcProvider.drawString(text, 0f, 0f, color)

            GL11.glPopMatrix()
        }
    }

    protected fun getStringWidth(fontSize: Int, name: String): Float {
        val font = FPSMaster.fontManager.getFont(fontSize)

        return if (mod.betterFont.value) {
            font.getStringWidth(name).toFloat()
        } else {
            ProviderManager.mcProvider.getFontRenderer().getStringWidth(name).toFloat()
        }
    }
}
